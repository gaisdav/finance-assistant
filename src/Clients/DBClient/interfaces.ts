import { DBSchema } from "idb";

export interface IDB extends DBSchema {
  amountStore: {
    key: string;
    value: number;
  };
}

export type TStoreNames = keyof Pick<IDB, "amountStore">;

export interface IDBClientStoragesConfig {
  storageName: TStoreNames;
  defaultValues?: [
    {
      [key: string]: any;
    }
  ];
}
