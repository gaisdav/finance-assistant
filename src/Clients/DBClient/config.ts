import { IDBClientStoragesConfig } from "./interfaces";

const idbStoragesConfigs: IDBClientStoragesConfig[] = [
  { storageName: "amountStore" },
];

export default idbStoragesConfigs;
