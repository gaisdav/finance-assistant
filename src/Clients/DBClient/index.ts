import { deleteDB, openDB } from "idb";
import { IDBPDatabase, StoreKey, StoreValue } from "idb/build/esm/entry";
import { IDB, IDBClientStoragesConfig, TStoreNames } from "./interfaces";

class DBClient {
  db: IDBPDatabase<IDB> | null = null;
  private DBName: string = "Finance Assistant";

  constructor(private configs: IDBClientStoragesConfig[]) {}

  async init() {
    this.db = await openDB<IDB>(this.DBName, 1, {
      upgrade: async (db) => {
        this.configs.forEach((storage) => {
          db.createObjectStore(storage.storageName);
        });
      },
    });
  }

  async get(
    storeName: TStoreNames,
    key: string
  ): Promise<StoreValue<IDB, TStoreNames> | null> {
    if (!this.db) {
      throw new Error("DB is not initialized");
    }

    return (await this.db.get(storeName, key)) || null;
  }

  async set(
    storeName: TStoreNames,
    key: string,
    val: any
  ): Promise<StoreKey<IDB, TStoreNames>> {
    if (!this.db) {
      throw new Error("DB is not initialized");
    }

    return this.db.put(storeName, val, key);
  }

  async delete(storeName: TStoreNames, key: string): Promise<void> {
    if (!this.db) {
      throw new Error("DB is not initialized");
    }

    await this.db.delete(storeName, key);
  }

  async dismiss() {
    await deleteDB(this.DBName, {
      blocked() {},
    });
  }
}

export default DBClient;
