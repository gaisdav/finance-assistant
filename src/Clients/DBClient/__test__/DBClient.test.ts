import DBClient from "../index";

const db = new DBClient();

describe("DBClient", () => {
  beforeAll(async () => {
    await db.init();
  });

  it("should DBClient", () => {
    db.set("asd", 4);
  });
});
