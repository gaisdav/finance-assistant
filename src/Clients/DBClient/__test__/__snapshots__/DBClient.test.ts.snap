// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DBClient снимок 1`] = `
Object {
  "_closePending": false,
  "_closed": false,
  "_rawDatabase": Database {
    "connections": Array [
      FDBDatabase {
        "_closePending": false,
        "_closed": false,
        "_rawDatabase": [Circular],
        "_runningVersionchangeTransaction": false,
        "listeners": Array [],
        "name": "Finance Assistant",
        "objectStoreNames": Array [
          "amountStore",
        ],
        "version": 1,
      },
    ],
    "deletePending": false,
    "name": "Finance Assistant",
    "processTransactions": [Function],
    "rawObjectStores": Map {
      "amountStore" => ObjectStore {
        "autoIncrement": false,
        "deleted": false,
        "keyGenerator": null,
        "keyPath": null,
        "name": "amountStore",
        "rawDatabase": [Circular],
        "rawIndexes": Map {},
        "records": RecordStore {
          "records": Array [],
        },
      },
    },
    "transactions": Array [
      FDBTransaction {
        "_objectStoresCache": Map {
          "amountStore" => FDBObjectStore {
            "_indexesCache": Map {},
            "_name": "amountStore",
            "_rawObjectStore": ObjectStore {
              "autoIncrement": false,
              "deleted": false,
              "keyGenerator": null,
              "keyPath": null,
              "name": "amountStore",
              "rawDatabase": [Circular],
              "rawIndexes": Map {},
              "records": RecordStore {
                "records": Array [],
              },
            },
            "autoIncrement": false,
            "indexNames": Array [],
            "keyPath": null,
            "transaction": [Circular],
          },
        },
        "_requests": Array [],
        "_rollbackLog": Array [
          [Function],
          [Function],
        ],
        "_scope": Set {
          "amountStore",
        },
        "_started": true,
        "_state": "finished",
        "db": FDBDatabase {
          "_closePending": false,
          "_closed": false,
          "_rawDatabase": [Circular],
          "_runningVersionchangeTransaction": false,
          "listeners": Array [],
          "name": "Finance Assistant",
          "objectStoreNames": Array [
            "amountStore",
          ],
          "version": 1,
        },
        "error": null,
        "listeners": Array [
          Object {
            "callback": [Function],
            "capture": false,
            "type": "error",
          },
          Object {
            "callback": [Function],
            "capture": false,
            "type": "abort",
          },
          Object {
            "callback": [Function],
            "capture": false,
            "type": "complete",
          },
          Object {
            "callback": [Function],
            "capture": false,
            "type": "complete",
          },
          Object {
            "callback": [Function],
            "capture": false,
            "type": "abort",
          },
        ],
        "mode": "versionchange",
        "objectStoreNames": Array [
          "amountStore",
        ],
        "onabort": null,
        "oncomplete": null,
        "onerror": null,
      },
    ],
    "version": 1,
  },
  "_runningVersionchangeTransaction": false,
  "listeners": Array [],
  "name": "Finance Assistant",
  "objectStoreNames": Array [
    "amountStore",
  ],
  "version": 1,
}
`;
